src/context/fragments-context.tsx

```diff
@@ -98,167 +28,156 @@ export const useFragment = (id?: string) => {
+				currentId: null,
+				loaded
 			});
-			// Adds any fragments in the payload which do not match any id in the current state's fragments.
-			const updatedFragments = mergedFragments
-				.concat(action.data.filter((actionFragment: any) => (
-					state.fragments.every((fragment: any) => fragment.id !== actionFragment.id)
-				)));
-			return {
-				...state,
-				fragments: updatedFragments,
-				loaded: action.loaded,
-				currentId: null
-			};
-		}
-		case FragmentActionType.UPDATE_ONE:
-			return updateOne(state, action);
-		case FragmentActionType.REMOVE_FRAGMENT: {
-			return {
-				...state,
-				fragments: state.fragments.filter((fragment: any) => fragment.id !== action.id),
-				loaded: action.loaded,
-				currentId: action.id
-			};
-		}
-		case FragmentActionType.TOGGLE_VISIBILITY: {
-			const fragmentToHide = state.fragments.find((fragment: any) => fragment.id === action.id);
-			fragmentToHide.hidden = action.hidden;
-			return {
-				...state,
-				loaded: action.loaded,
-				currentId: action.id
-			};
-		}
-		case FragmentActionType.REMOVE_FRAGMENTS: {
-			const remainingFragments = state.fragments.filter((fragment: any) => (
-				!action.ids.some((actionFragment: any) => actionFragment.id === fragment.id)
-			));
-			return {
-				...state,
-				fragments: remainingFragments,
-				loaded: action.loaded,
-				currentId: null
-			};
-		}
-		case FragmentActionType.ADD_ONE: {
-			const duplicate = assign({}, action.data);
-			const expandedFragments = state.fragments.concat(duplicate);
-			return {
-				...state,
-				fragments: expandedFragments,
-				loaded: action.loaded,
-				currentId: action.data.id
-			};
+			return;
 		}
-		default:
-			return state;
+
+		// Remove fragments which are in the original state but not in the payload.
+		const filteredFragments = fragmentsState.fragments
+			.filter((fragment: any) => fs.fragments.some((actionFragment: any) => actionFragment.id === fragment.id));
+		// If fragments already exist in the state, we need to merge any changes to the fragments with
+		// the current matching fragments and add any new fragments (if any) to the state.
+		const mergedFragments = filteredFragments.map((fragment: any) => {
+		// Find the fragment in the payload containing the same id as the current state's fragments
+		// to merge updated changes with.
+			const updatedFragment = fs.fragments.find((actionFragment: any) => fragment.id === actionFragment.id);
+			// Can not use merge because removing datasets or labels will not
+			// work since it keeps the values, while assign overwrites past values.
+			return assign({}, fragment, updatedFragment);
+		});
+		// Adds any fragments in the payload which do not match any id in the current state's fragments.
+		const updatedFragments = mergedFragments
+			.concat(fs.fragments.filter((actionFragment: any) => (
+				fragmentsState.fragments.every((fragment: any) => fragment.id !== actionFragment.id)
+			)));
+
+		_setFragmentsState({
```

Comment:
I might be reading this wrong, but in case I'm not... calling the `useState`'s set function twice in a row and expecting the state (returned by that useState) to be updated between them won't go well...
First call was in the if statement above.

Reply:
Also, at what point is the state saved to localStorage?

Reply:
> I might be reading this wrong, but in case I'm not... calling the `useState`'s set function twice in a row and expecting the state (returned by that useState) to be updated between them won't go well... First call was in the if statement above.

good catch, the first call should have had a return at the end

Reply:
> Also, at what point is the state saved to localStorage?

In the use effect right before returning the template:
```typescript
	useEffect(() => {
		// store only ids to local storage so we don't get into temptation of using other
		// props that should really be coming from db
		localStorage.setItem('localFragments', JSON.stringify(fragmentsState.fragments));
	}, [fragmentsState.fragments]);
```

I didn't make any changes there

Reply:
> It returns if that first one was called

ohh, after a 50 line red block... 🤦